mava/networks/utils/mat/decode.py

```diff
@@ -0,0 +1,171 @@
+# Copyright 2022 InstaDeep Ltd. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import Tuple, Union
+
+import chex
+import jax
+import jax.numpy as jnp
+import tensorflow_probability.substrates.jax.distributions as tfd
+from flax import linen as nn
+
+from mava.networks.distributions import IdentityTransformation, TanhTransformedDistribution
+
+# General shapes legend:
+# B: batch size
+# N: number of agents
+# O: observation dimension
+# A: action dimension
+# E: model embedding dimension
+
+
+def discrete_parallel_act(
+    decoder: nn.Module,
+    obs_rep: chex.Array,  # (B, N, E)
+    action: chex.Array,  # (B, N)
+    action_dim: int,  # (, )
+    legal_actions: chex.Array,  # (B, N, A)
+    key: chex.PRNGKey,
+) -> Tuple[chex.Array, chex.Array]:
+    batch_size, n_agent, _ = obs_rep.shape
+    one_hot_action = jax.nn.one_hot(action, action_dim)  # (B, A)
+    shifted_action = jnp.zeros((batch_size, n_agent, action_dim + 1))  # (B, N, A +1)
+    shifted_action = shifted_action.at[:, 0, 0].set(1)
+    shifted_action = shifted_action.at[:, 1:, 1:].set(one_hot_action[:, :-1, :])
+    logit = decoder(shifted_action, obs_rep)  # (B, N, A)
```

Comment:
It goes a bit beyond just star tof sequence tokens or not and it is also not really the parallel versions that are affected. It is because when we use heads we cannot pass in a single agent's embedded observation at a time nor can we mask a single agents actions when initialising the distributions for action selection. When using heads we are forced to pass all the agent's observations and legal action masks through at the same time. This had a non-zero effect on the performance when I did it to test if using heads was possible and for this reason I think this is something we should skip for now. Make and issue and consider looking at in the future. 

I don't see this as a major issue since these transformer based systems are already so different from other systems we have in Mava that being slightly different here for the sake of performance is okay in my opinion.